### PR TITLE
Add the `final` keyword to all shipped classes

### LINKED ISSUE #46: Introduce final keyword where appropriate
### Feature Request

Most classes here have been marked `@final` so in `4.0.0` we can introduce the `final` keyword throughout

### PR DESCRIPTION
**Pull request by gsteel**
|    Q          |   A
|-------------- | ------
| BC Break      | yes
| QA            | yes

### CODE REVIEW

File: test/Asset/FixedResponseMiddleware.php [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+declare(strict_types=1);
+
+namespace MezzioTest\Router\Asset;
+
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+final class FixedResponseMiddleware implements MiddlewareInterface
```
Comment (lcobucci) [inline test/Asset/FixedResponseMiddleware.php:12]: It's quite clear to me that there's a lot of value in using components like this to simplify tests.

Is there a reason to not offer them to users (aka keep them in the "test")?

**Review by lcobucci — APPROVED** [review]
LGTM

File: test/Asset/FixedResponseMiddleware.php (hunk 1 above)
Comment (gsteel) [reply test/Asset/FixedResponseMiddleware.php:12]: I'm not sure I follow - do you mean ship a separate package that contains all the usual stuff like 'InMemoryContainer' and PSR-related testing tools? I always consider anything in /test to be void of any BC gurantees, but a separate package for require-dev would save some repetition…

Comment (lcobucci) [reply test/Asset/FixedResponseMiddleware.php:12]: I don't think we necessarily need to have different packages, honestly. We'd probably only increase our maintenance cost. 

The likelihood of having changes in these components is small, and they bring value to the ecosystem by promoting better practices in testing (though their applicability isn't limited to that scope).

So, my suggestion is to move them to the source folder.

Comment (gsteel) [reply test/Asset/FixedResponseMiddleware.php:12]: I think it's a good idea - The amount of duplicated tooling I have like this in projects is considerable, but, it's a maintenance burden anywhere we ship it so I'm not totally convinced 🤔

Comment (lcobucci) [reply test/Asset/FixedResponseMiddleware.php:12]: I understand.

The components are a maintenance burden as long as they exist, moving them to source just increases the burden 😅.

I'm good either way, honestly, just think it's an opportunity to bring value with  little cost - based on what I see in other projects, like: Monolog (`NullHandler` and `TestHandler`) and EventSauce (`CollectingMessageConsumer` and other `CollectingMessage*`).

**Pull request merged**